Grant/contracts/TipperGrant.sol

```diff
@@ -0,0 +1,174 @@
+        uint24 claimedCount;
+        uint112 totalGrantAmount;
+        uint112 claimedGrantAmount;
+        bytes32 merkleRoot;
+        mapping(address => uint112) claimedAmount; // Track claimed amount per address
+    }
+    Season[] private seasons;
+
+    event Deposit(address indexed user, uint256 huntAmount);
+    event SetGrantData(uint256 indexed season, uint24 walletCount, bytes32 merkleRoot);
+    event Claim(address indexed user, uint256 indexed season, uint112 huntAmount);
+    event EmergencyWithdraw(address indexed user, uint256 huntAmount);
+
+    function currentSeason() external view returns (uint256) {
+        return seasons.length;
+    }
+
+    function deposit(uint256 huntAmount) external {
+        address msgSender = _msgSender();
+        if (!HUNT.transferFrom(msgSender, address(this), huntAmount)) revert TokenTransferFailed();
+
+        emit Deposit(msgSender, huntAmount);
+    }
+
+    function emergencyWithdraw() external onlyOwner {
+        address msgSender = _msgSender();
+        uint256 balance = HUNT.balanceOf(address(this));
+        if (!HUNT.transfer(msgSender, balance)) revert TokenTransferFailed();
+
+        emit EmergencyWithdraw(msgSender, balance);
+    }
+
+    modifier validSeasonId(uint256 seasonId) {
+        if (seasonId >= seasons.length) revert InvalidSeasonId();
+        _;
+    }
+
+    function setGrantData(
+        uint256 seasonId,
+        uint24 walletCount,
+        uint112 totalGrant,
+        bytes32 _merkleRoot
+    ) external onlyOwner {
+        Season storage season;
+        if (seasonId == seasons.length) {
+            seasons.push();
+            season = seasons[seasonId];
+        } else if (seasonId < seasons.length) {
+            // overwrites existing season data, unless anyone has already claimed
+            season = seasons[seasonId];
+            if (season.claimedGrantAmount > 0) revert SeasonDataCannotBeUpdated();
+        } else {
+            revert InvalidSeasonId();
+        }
+        if (totalGrant == 0) revert InvalidDepositAmount();
+        if (walletCount == 0) revert InvalidWalletCount();
+        if (totalGrant > HUNT.balanceOf(address(this))) revert NotEnoughGrantBalance();
+
+        season.walletCount = walletCount;
+        season.totalGrantAmount = totalGrant;
+        season.merkleRoot = _merkleRoot;
+        season.claimStartedAt = uint40(block.timestamp);
+
+        emit SetGrantData(seasonId, walletCount, _merkleRoot);
+    }
+
+    function claim(uint256 seasonId, uint112 amount, bytes32[] calldata merkleProof) external validSeasonId(seasonId) {
+        Season storage season = seasons[seasonId];
+        address msgSender = _msgSender();
+
+        if (block.timestamp > season.claimStartedAt + CLAIM_DEADLINE) revert ClaimDeadlineReached();
+        if (season.claimedAmount[msgSender] > 0) revert AlreadyClaimed();
+        if (!_verify(season.merkleRoot, msgSender, amount, merkleProof)) revert InvalidMerkleProof();
```

Comment:
I don't have enough time to confirm this but don't you have to verify the merkleProof length > 0?

Reply:
We have some test cases (https://github.com/Steemhunt/hunt-town/pull/8/files#diff-b901f3036186c5aa75f32d71405285ef109fcc72d84cedc3171c47bb2436bdf8R269
) for empty merkleProof, but it might be safer to check it again because we don’t expect single-node trees in any case.